### ISSUE TITLE
some better path approach for executables besides stack exec

### ISSUE THREAD

**Issue by wolftune** [commercialhaskell/stack#1007]
I'll try to describe the issue clearly.

* `stack install ghc-mod` works to install ghc-mod system-wide, and as long as on path, it will work with text-editor plugins that use ghc-mod.
    * incidentally, contrary to the GUIDE instructions to use .bashrc for setting path, the path has to be set in bash_profile for text editors like Atom that may get instantiated outside of terminal. Alternatively, the haskell-ghc-mod Atom package has an option to specify the path
* However, what if someone has multiple projects that use different GHC versions. Stack enables this. But now, a single ghc-mod build is not workable for both as it should use the same GHC as the project.
* If we stick to `stack build` and use `stack exec` to run ghc-mod, it won't work for integration with ghc-mod based editor tools
* So, there needs to be *some* reasonable way to specify path to the local .stack-work for running executables without using `stack exec`

With cabal sandboxes, it worked to add .cabal-sandbox/bin to the .bashrc path, and then anything built locally in a project could be accessed if run from the project. Cabal wouldn't support this exact situation, but in general, it was possible to get the executable on the path without an extra command and still be project-specific.

I really am not sure what the overall solution is here with so many layers, but the hiding of project-specific executables behind `stack exec` is definitely an issue in this case.

**Comment by mgsloan** [comment]
`stack path` should give you the info needed here.  If it doesn't already, then it should!

`stack path --snapshot-install-root` gives the path to the binary created by `stack build ghc-mod`.  

It might also be worth first looking in `stack path --local-install-root`, which is where binaries from the current project go.

**Comment by wolftune** [comment]
@mgsloan Thanks! Incidentally, ghc-mod isn't in an LTS snapshot yet, although it's in nightly… but those commands should be enough to work around the issues here somewhat.

It still doesn't address the full details of a situation where someone wanted ghc-mod editor integration with multiple projects and different GHC versions, but that's something that could be fixed in the editor packages.

**Comment by wolftune** [comment]
Note, I'm opening a relevant [ticket](https://github.com/atom-haskell/haskell-ghc-mod/issues/54) for the Atom haskell-ghc-mod package.

**Comment by wolftune** [comment]
Okay, re-opening here. The issue is basically that it is burdensome to deal with extra support from all sorts of third-party tools if we could get a generalized path.

It works with cabal sandboxes to just have `.cabal-sandbox/bin` on the $PATH. Then, the sandboxed executables will be available based on whatever project one is in, and because no extra command like `stack exec` is needed, any tool that just calls an executable wouldn't have to be updated to deal with the varying paths within .stack-work or to use stack exec.

So, how about a consistent, approved stack command to copy executables to a *local* project directory that's generalizable across projects? That could remove the need for `stack exec` and allow having *local* executables on a shell-recognized path.

In summary, that's really the issue: there's value in having an executable on a regular path, but there's also value in having a path that always uses the local version for each project.

**Comment by wolftune** [comment]
One more clarifying thought: If I set a path in a tool after getting it from `stack path` (or `stack exec where ghc-mod`), I still have to *manually* update that path every time the project updates to a new LTS version.

So, the short version is: **we need some way to specify a path for tools to access stack-built executables that is *local* to each project but persistent across LTS updates**.

**Comment by snoyberg** [comment]
I'm still having a hard time understanding why the existing `stack path`, `stack exec env`, and `stack exec` commands don't meet this need.

**Comment by wolftune** [comment]
@snoyberg I don't think this is a big deal or high priority, it's just this: with cabal sandboxes, one could put .cabal-sandbox on the PATH, and that would mean that *if* run from the project directory, any arbitrary tool will see whatever those items in the PATH in the sandbox are. Thus, those third-party executables don't need to know anything about cabal or sandboxes or how anyone has their system set up.

In the case of stack, I know of now way to reliably put the .stack-work executables on the PATH when in the project root directory. Thus, the user or the developers of various tools have to do something with a stack command or a setting of the PATH for the tool or other such workarounds. In other words, the work flow is forced to adapt to Stack instead of just say "as long as it's on the PATH".

I don't know how many tools are affected in practice, but in principle, it just would allow maximum flexibility and not offload the concerns to users and third-party tools *if* there were a way to have a PATH wildcard sort of setting that would find any .stack-work executable regardless of where in the .stack-work subdirectories it is. If in fact, there's a PATH setting that would do that, it would be a complete solution… alternatively, I could see a `stack install local` or something that would move executables not to a system-wide PATH but just to `.stack-work/bin` so that could be put on the PATH and then executables could be available without the workarounds and still stay only in the scope of each project.

**Comment by snoyberg** [comment]
This only works with sandboxes when you run with `cabal install`, not `cabal build`, and then don't change your GHC version in between. That seems like exactly what `stack install` is doing to be honest.

Adding yet another directory where executables get copied to - assuming the user uses one command instead of another - doesn't seem like it will make the situation better, only worse.

Look at the output of `stack exec env | grep PATH`, that's probably exactly the information you're looking for, and will support cases such as changing between different LTS versions or compilers.

**Comment by wolftune** [comment]
Yes, to clarify, I was saying that since `cabal install` (versus build) works with sandboxes, I could see cases where a comparable *sandboxed* install with Stack is useful. If you disagree, you can reject the idea and close this. I don't quite see anything else as an identical solution, so I'd like to see the option available.

**Comment by snoyberg** [comment]
You can use `stack install --local-bin-path bin` to install executables into a local directory. But they won't actually be properly sandboxed; if you later change your stack.yaml file, those old executables will still be used (which is the same problem with cabal sandboxes).

**Comment by wolftune** [comment]
okay, sounds like there's just no straightforward solution at this time. Third-party tools that need to call truly sandboxed executables will just need to use Stack-specific options recognizing the `stack exec` commands or the stack-work internal paths. Obviously lots of other challenges to deal with besides this one. Thanks for your time thinking about it.
